### DISCUSSION TITLE
RPN mathematics scratchpad

### DISCUSSION THREAD

**Discussion by ajb96817** [KaTeX/KaTeX#4005] [Show and tell]
I've been working on this math expression editor for personal use for some time.  It's now fairly complete.  It's designed for keyboard-only operation and uses a RPN (stack-based) entry system.  It uses React and KaTeX and should work in any browser.

Because of all the specialized keyboard bindings there's a learning curve, but I've found it a very efficient way of entering formulas and working problems as a pen-and-paper alternative.

  * Source: https://github.com/ajb96817/rpnlatex/
  * Try it out at: https://ajb96817.github.io/rpnlatex/

**Comment by tail-call** [comment]
Hi @ajb96817!

I tried the scratchpad you made. I attempted to enter `$w(z) = e^{-z^2}\mathrm{erfc}(-iz)$` and gave up somewhere trying to get erfc to appear, but it was a fun ride!

What I liked:
- I imagine it's *very* convenient when you learn all the commands
- The user guide is very comprehensive
- Vi keybindings are an instant 🏅 from me
- Everything happens instantly

What could be improved:
- Ctrl+Z works, but please also add support for Cmd+Z on mac :)
- Pasting from buffer (Ctrl+V/Cmd+V) would also be very convenient
- A tutorial explaining the whole process from starting the scratchpad to getting the final output would really help. I found that entering Tab-$ produces LaTeX source, but I couldn't copy it
- Entering multi-digits number now, it seems, requires typing the entire number, then joining digits by pressing space many times. I couldn't figure out how to enter decimal fraction
- When I press Backspace, my whole expression is gone. This is counterintuitive, because when editing text, I'm not used to having my entire input erased when I press Backspace

Thank you for sharing this. I look forward to seeing how it evolves in the future. Keep up the great work!

**Comment by ajb96817** [comment]
Thanks so much for trying it!  For your example, you can enter \mathrm{erfc} via the \ prefix, then type erfc, then finish with Shift+Enter which will typeset in \mathrm.  Then enter -iz (i z Space . -) and apply with /o or Ctrl+O.

So for your example the keystrokes would be:

w z / o       -> w(z)
z . 2 . - / e    -> e^(-z^2)
\ e r f c Shift+Enter  -> erfc
i z Space . -    -> -iz
/ o   -> erfc(-iz)
Space  -> e^(-z^2) erfc(-iz)
= =   -> final expression

It does take some getting used to I admit!

I'll see about Cmd+Z, I don't have a Mac so not sure how Cmd keys appear to the browser.

Pasting LaTeX code from the system clipboard would be a good feature, I agree.  There is also an internal clipboard available with Tab c and Tab v but those operate directly on the math items, not the LaTeX source.

You can copy (export) the stack top item to the system clipboard as LaTeX via Tab @ or the entire document with Tab !

I'm planning on putting in a proper export-to-latex option that generates something you can feed directly to pdflatex, etc, but it's not quite there yet.  Tab ! can get you most of the way except for the LaTeX prologue.

For multi-digit numbers and decimals you can use the \ prefix, e.g. \ 1 2 3 . 4 5 Enter, it supports fractions too: \ 1/2 Enter

Backspace is meant to drop the entire item from the stack but for simple-enough expressions you can use Shift+Enter to edit the stack top as text (this is very limited at the time though).  This program doesn't exactly parse LaTeX directly, instead it builds up expressions into a tree-structure as you go, so instead of editing the LaTeX code itself you could use the destructuring commands such as / < (to get the left side of an expression) or use the "dissect" mode available with _ (underscore).

Thanks again for your feedback!

- Andrew
